client/data/site-settings/with-site-settings.js

```diff
@@ -0,0 +1,151 @@
+			},
+			[ dispatch ]
+		);
+
+		const trackTracksEvent = useCallback(
+			( name, eventProps ) => {
+				dispatch( recordTracksEvent( name, eventProps ) );
+			},
+			[ dispatch ]
+		);
+
+		const { data, isLoading: isFetchingSettings } = useSiteSettingsQuery( siteId );
+
+		const updateSettings = useCallback(
+			( fields ) => {
+				markChanged();
+				return setUnsavedSettings( { ...unsavedSettings, ...fields } );
+			},
+			[ markChanged, unsavedSettings, setUnsavedSettings ]
+		);
+
+		const {
+			isPending: isSavingSettings,
+			updateSiteSettings: saveSettings,
+			context: mutateContext,
+		} = useUpdateSiteSettingsMutation( siteId, {
+			onMutate: async ( settings ) => {
+				const queryKey = [ 'site-settings', siteId ];
+
+				// Cancel any current refetches, so they don't overwrite our optimistic update
+				await queryClient.cancelQueries( {
+					queryKey,
+				} );
+
+				// Snapshot the previous value
+				const previousSettings = queryClient.getQueryData( queryKey );
+
+				// Optimistically update to the new value
+				queryClient.setQueryData( queryKey, ( { settings: oldSettings } ) => {
+					return { ...oldSettings, ...settings };
+				} );
+
+				// Store previous settings in case of failure
+				return { previousSettings };
+			},
+			onSuccess() {
+				dispatch( successNotice( translate( 'Settings saved successfully!' ), noticeOptions ) );
+				setUnsavedSettings( {} ); // clear dirty fields after successful save.
+				if ( path === '/settings/newsletter/:site' ) {
+					trackTracksEvent( 'calypso_settings_newsletter_saved', unsavedSettings );
+				}
+				markSaved();
+			},
+			onError( err, newSettings, context ) {
+				// Revert to previous settings on failure
+				queryClient.setQueryData( [ 'site-settings', siteId ], context.previousSettings );
+
+				dispatch(
+					errorNotice(
+						translate( 'There was a problem saving your changes. Please, try again.' ),
+						noticeOptions
+					)
+				);
+			},
+			onSettled: () => {
+				// Refetch settings regardless
+				queryClient.invalidateQueries( {
+					queryKey: [ 'site-settings', siteId ],
+				} );
+			},
+		} );
+
+		const handleSubmitForm = useCallback( () => {
+			trackEvent( 'Clicked Save Settings Button' );
+			saveSettings( unsavedSettings );
+			if ( ! unsavedSettings ) {
+				return;
+			}
+			trackEvent( 'Clicked Save Settings Button' );
+		}, [ saveSettings, unsavedSettings, trackEvent ] );
+
+		const handleToggle = useCallback(
+			( name ) => () => {
+				if ( unsavedSettings[ name ] === undefined ) {
+					unsavedSettings[ name ] = data?.settings[ name ];
+				}
+				updateSettings( { [ name ]: ! unsavedSettings[ name ] } );
+				trackEvent( `Toggled ${ name }` );
+			},
+			[ trackEvent, unsavedSettings, updateSettings, data?.settings ]
+		);
+
+		const settings = Object.assign(
+			{},
+			mutateContext?.previousSettings?.settings,
+			data?.settings,
+			unsavedSettings
+		);
+		return (
+			<Wrapped
```

Comment:
If we're doing this "the new way", it might be more convenient to just use a hook instead of creating a HoC.